cpp/ql/src/semmle/code/cpp/controlflow/IRGuards.qll

```diff
@@ -405,20 +406,37 @@ class IRGuardCondition extends Instruction {
    */
   private predicate controlsBlock(IRBlock controlled, boolean testIsTrue) {
     not isUnreachedBlock(controlled) and
-    exists(IRBlock branchBlock | branchBlock.getAnInstruction() = branch |
-      exists(IRBlock succ |
-        testIsTrue = true and succ.getFirstInstruction() = branch.getTrueSuccessor()
-        or
-        testIsTrue = false and succ.getFirstInstruction() = branch.getFalseSuccessor()
-      |
-        branch.getCondition() = this and
-        succ.dominates(controlled) and
-        forall(IRBlock pred | pred.getASuccessor() = succ |
-          pred = branchBlock or succ.dominates(pred) or not pred.isReachableFromFunctionEntry()
-        )
+    // The following implementation is ported from `controls` in Java's
+    // `Guards.qll`. See that file (at 398678a28) for further explanation and
+    // correctness arguments.
+    exists(IRBlock succ |
+      this.hasBranchEdge(succ, testIsTrue) and
+      this.hasDominatingEdgeTo(succ) and
+      succ.dominates(controlled)
+    )
+  }
+
+  /**
+   * Holds if `(this, succ)` is an edge that dominates `succ`, that is, all other
+   * predecessors of `succ` are dominated by `succ`. This implies that `this` is the
+   * immediate dominator of `succ`.
+   *
+   * This is a necessary and sufficient condition for an edge to dominate anything,
+   * and in particular `bb1.hasDominatingEdgeTo(bb2) and bb2.dominates(bb3)` means
+   * that the edge `(bb1, bb2)` dominates `bb3`.
+   */
+  private predicate hasDominatingEdgeTo(IRBlock succ) {
+    exists(IRBlock branchBlock | branchBlock = this.getBranchBlock() |
+      branchBlock.immediatelyDominates(succ) and
+      branchBlock.getASuccessor() = succ and
+      forall(IRBlock pred | pred = succ.getAPredecessor() and pred != branchBlock |
```

Comment:
This drops the requirement that `pred` is reachable. I think that's now a consistency requirement, but if it's violated, would that lead to a tuple explosion?

Reply:
Well spotted. I didn't drop the requirement on purpose, but I'd like to understand whether it's needed.

Is the problem that an unreachable `pred` is in principle dominated by every node because [the definition of dominance](https://en.wikipedia.org/wiki/Dominator_(graph_theory)) says that "`pred` is dominated by `succ` if for every path to `pred` [...]", and this holds vacuously because there are no paths to `pred`?

Does our `dominates` predicate actually implement that definition? We implement it as `immediatelyDominates*`, where `immediatelyDominates` is computed by a HOP. I expect that `immediatelyDominates` does not hold for unreachable nodes since otherwise it wouldn't be a tree. Therefore our `dominates` relation shouldn't hold for unreachable nodes either.

Reply:
That leads to some false negatives, then. If a `pred` is unreachable, then `succ.dominates(pred)` won't hold, and so the whole `forall` will fail. We do currently have at least one situation where the exit block of a function with an unconditional `throw` is kept in later stages, but that wouldn't trigger this issue (and might be a bug in any case).

@dbartol Should unreachable blocks be a consistency violation? I'm happy to merge this if so, otherwise I'd like to re-add the `not pred.isReachableFromFunctionEntry()`.

Reply:
I buy your argument, so I've restored the case along with a comment for Future Me.

If `isReachableFromFunctionEntry` always holds, then I suggest removing that predicate entirely and documenting on `class IRBlock` that a block is always reachable from the entry point. But that's outside the scope of this PR.